File: calculateDice.m
function dice = calculateDice(VOI1,VOI2,struct_num_1,struct_num_2)
% calculateDice calculates the Dice coefficient between two contours in
% terms of overlapping volume. 
%
% The build-in Matlab funcion bitget is used to get the correct contour
% within the VOI. 
% 
% dice = 2 |X ? Y|      using the volumes
%        ----------
%        |X| + |Y|   
%
% Inputs:       2 VOIs          2 volumes of interest of two structure sets. These VOIs include all 
%                               OARs/structures available in the RTSTRUCT that was used as input    
%               struct_num_1    the number of the OAR/structure within VOI1
%               struct_num_2    the number of the OAR/structure within VOI2
%
% Outputs:      dice            the dice similarity coefficient, represented as a fraction
%
%
%     Femke Vaassen @ MAASTRO.


disp('-   Calculating: Dice similarity coefficient');

% Taking the structure that is wanted (e.g. the heart) out of the VOI using bitget
VOI1 = bitget(VOI1,struct_num_1); 
VOI1 = double(VOI1);

VOI2 = bitget(VOI2,struct_num_2);
VOI2 = double(VOI2);

% Calculating the overlapping pixels and using that to calculate the dice
% similarity coefficient.
pixelDataOverlap = VOI1 & VOI2; 
dice = (2*sum(pixelDataOverlap(:))) / (sum(VOI1(:))+ sum(VOI2(:))); 

end

File: calculate_dice.py
"""
calculateDice - Calculate Dice coefficient between two contours

Calculates the Dice coefficient between two contours in terms of overlapping volume.

dice = 2 |X ∩ Y|      using the volumes
       ----------
       |X| + |Y|

Femke Vaassen @ MAASTRO
"""

import numpy as np


def calculate_dice(voi1, voi2, struct_num_1, struct_num_2):
    """
    Calculate the Dice similarity coefficient between two structures.
    
    Parameters
    ----------
    voi1 : ndarray
        First volume of interest (contains all structures)
    voi2 : ndarray
        Second volume of interest (contains all structures)
    struct_num_1 : int
        Structure number (1-indexed) within VOI1
    struct_num_2 : int
        Structure number (1-indexed) within VOI2
        
    Returns
    -------
    float
        Dice similarity coefficient (0 to 1)
    """
    print('-   Calculating: Dice similarity coefficient')
    
    # Extract the specific structure using bit operations
    # MATLAB bitget is 1-indexed, we need to adjust for 0-indexed Python
    # bitget(voi1, struct_num_1) extracts bit at position struct_num_1
    voi1_struct = ((voi1 >> (struct_num_1 - 1)) & 1).astype(bool)
    voi2_struct = ((voi2 >> (struct_num_2 - 1)) & 1).astype(bool)
    
    # Calculate overlapping pixels
    pixel_data_overlap = voi1_struct & voi2_struct
    
    # Calculate Dice coefficient
    dice = (2.0 * np.sum(pixel_data_overlap)) / (np.sum(voi1_struct) + np.sum(voi2_struct))
    
    return dice
